### PR TITLE
Improvements to DragForcing, TerrainDrag, ChannelBuilder

### LINKED ISSUE #1969: Make modifications so wall drag can be applied above or below
- [ ] changes to ChannelBuilder to flag drag cells
- [ ] changes to DragForcing to look above or below to calculate terrain drag quantities
- [ ] possibly the type of changes for Kosovic (looking above or below)

### LINKED ISSUE #1970: Modify DragForcing to avoid using strange conditionals related to cell size and is_laminar
For estuary flows validation efforts

### PR DESCRIPTION
**Pull request by mbkuhn**
## Summary

- Allow user to specify some previously hard-coded constants and override default behavior. closes #1970
- Enable TerrainDrag to identify drag cells below blanked terrain cells (not just above them) and make DragForcing smart enough to handle that. closes #1969 
- Enable ChannelBuilder to label necessary fields for wall model in DragForcing
- Enable ChannelBuilder to turn off velocity modification

So far, this doesn't include the option of using a smooth wall model.

## Pull request type

Please check the type of change introduced:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Checklist

The following is included:

- [ ] new unit-test(s)
- [x] new regression test(s)
- [ ] documentation for new capability

This PR was tested by running:

- the unit tests
  - [ ] on GPU 
  - [ ] on CPU 
- the regression tests
  - [ ] on GPU 
  - [ ] on CPU 

## Additional background

Issue Number:

### CODE REVIEW

File: src/equation_systems/icns/source_terms/DragForcing.cpp [hunk 1]
```diff
@@ -258,13 +261,14 @@ void DragForcing::operator()(
...
+    const amrex::Real min_z0 = m_min_z0;
+    const amrex::Real scale_factor =
+        (m_limit_drag && dx[2] < 1.0_rt) ? 1.0_rt : 1.0_rt / dx[2];
+    const amrex::Real Cd = (m_limit_drag && is_laminar != 0 && dx[2] < 1)
                                ? drag_coefficient
                                : drag_coefficient / dx[2];
     const amrex::Real kappa = m_kappa;
-    const amrex::Real z0_min = 1.0e-4_rt;
-    const amrex::Real cd_max = 1000.0_rt;
+    const amrex::Real cd_max = m_limit_drag ? m_cd_max : constants::LARGE_NUM;
```
Comment (hgopalan) [inline src/equation_systems/icns/source_terms/DragForcing.cpp:273]: A large number can make the system very stiff and unstable. The limit of 1000 was based on the analysis in FastEddy. We may have a higher number but should still restrict.

Comment (mbkuhn) [reply src/equation_systems/icns/source_terms/DragForcing.cpp:273]: The default behavior is the same. But the user can turn it off if desired

File: test/test_files/channel_terrain_walls/channel_terrain_walls.inp [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨#
```
Comment (d-montgomery) [inline test/test_files/channel_terrain_walls/channel_terrain_walls.inp:1]: This is nit-picky, but should this be called `channel_builder_terrain_walls/channel_builder_terrain_walls.inp` so it's clear that it is using the `ChannelBuilder` physics?

File: test/test_files/channel_terrain_walls/channel_terrain_walls.inp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+amr.max_level           = 0           # Max AMR level in hierarchy
+
+#¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨#
+#              GEOMETRY                 #
+#.......................................#
+geometry.prob_lo        =   0.0 0.0 0.0  # Lo corner coordinates
+geometry.prob_hi        =   0.5 0.5 1.0668  # Hi corner coordinates 8*pi*3pi*1
+geometry.is_periodic    =   1   1  0   # Periodicity x y z (0/1)
+
+# Boundary conditions
```
Comment (d-montgomery) [inline test/test_files/channel_terrain_walls/channel_terrain_walls.inp:71]: I ran this test case and wondered if this checker-boarding in the z-velocity is expected because of the resolution, or if there is a larger problem here. The velocity in the first fluid cell is alternating between $\pm$ the tangental velocity of `zhi`.  This was from commit [33273b8](https://github.com/kynema/kynema-sgf/pull/1983/commits/33273b8e84ade8a9b6ed8646f3f1550363a1519b).

Image

**Review by d-montgomery — COMMENTED** [review]
Thanks for adding these capabilities, @mbkuhn. I ran the test case and am playing with it now in the validation tests I've been working on. I have a couple minor comments.

File: test/test_files/channel_terrain_walls/channel_terrain_walls.inp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+zlo.type = "slip_wall"
+zhi.type =   "no_slip_wall"
+zhi.velocity = 0.5646234542509774 0.0 0.0
```
Comment (mbkuhn) [inline test/test_files/channel_terrain_walls/channel_terrain_walls.inp]: remove this! leftover from other test

**Review by hgopalan — APPROVED** [review]

File: test/test_files/channel_terrain_walls/channel_terrain_walls.inp (hunk 2 above)
Comment (mbkuhn) [reply test/test_files/channel_terrain_walls/channel_terrain_walls.inp:1]: I used the name channel_terrain_walls to indicate that it is using the channel physics like the other channel cases, but the walls employ the terrain capability. If I used channel_builder_terrain_walls that would be somewhat redundant because the channel builder always uses terrain for walls. What about channel_builder_channel? Is that more clear?

File: test/test_files/channel_terrain_walls/channel_terrain_walls.inp (hunk 3 above)
Comment (mbkuhn) [reply test/test_files/channel_terrain_walls/channel_terrain_walls.inp:71]: DragForcing.use_temporal_limiter = true fixes this instability. The instability is caused by the z-component of DragForcing in drag cells, which ends up removing more momentum than is in the cell, causing it to switch sign and grow with time. With the default value of drag_coefficient = 10, this problem occurs when dt > 0.1 dx. Even though the same condition applies to the terrain cells, those are forced to a velocity magnitude of 0, diminishing the forcing term (which has a factor of m) there, whereas the velocity magnitude in drag cells will remain nonzero due to the parallel flow.

File: test/test_files/channel_terrain_walls/channel_terrain_walls.inp (hunk 2 above)
Comment (d-montgomery) [reply test/test_files/channel_terrain_walls/channel_terrain_walls.inp:1]: `channel_terrain_walls` is great.

Comment (d-montgomery) [reply test/test_files/channel_terrain_walls/channel_terrain_walls.inp:1]: Or whatever you want to name it!

File: docs/sphinx/user/inputs_Momentum_Sources.rst [hunk 5]
```diff
@@ -175,24 +175,36 @@ Section: Momentum Sources
 
 .. input_param:: DragForcing.drag_coefficient
 
-   **type:** Real, optional
+   **type:** Real, optional, default = 10.0
 
    This value specifies the coefficient for the forcing term in the immersed boundary forcing method. It is currently
-   recommended to use the default value to avoid initial numerical stability. 
+   recommended to use the default value to avoid initial numerical stability. Despite the name, the drag-related
```
Comment (d-montgomery) [inline docs/sphinx/user/inputs_Momentum_Sources.rst]: Avoid initial numerical "in"stability?

**Review by d-montgomery — COMMENTED** [review]
Great job resolving the checkerboarding issue, @mbkuhn! I have one minor comment then this is good to go.

File: docs/sphinx/user/inputs_Momentum_Sources.rst (hunk 5 above)
Comment (mbkuhn) [reply docs/sphinx/user/inputs_Momentum_Sources.rst]:
```suggestion
   recommended to use the default value to avoid initial numerical instability. Despite the name, the drag-related
```

**Review by d-montgomery — APPROVED** [review]
